GET /products, GET /reports and PUT /reports endpoints

BartekWons opened the pull request.

BartekWons committed:
add connection between tables resources and products table in /products endpoint

BartekWons committed:
add report endpoint

BartekWons committed:
add PUT /reports endpoint

BartekWons committed:
code touch-ups

vaexey commented on back/model/user.go:
```diff
@@ -1,12 +1,13 @@
+	DisplayName string   `json:"displayName"`
+	Password    string   `json:"-"`
+	IsAdmin     bool     `json:"-"`
+	AvatarId    *int     `json:"-"`
```
Inconsistent syntax for defining fk ID fields. For comparison: `model/contrib.go:10`

vaexey replied:
Issue has not been fixed properly.  Naming field `AvatarID` would not require adding the gorm `foreignKey` tag.

sebasion9 replied:
Gorm model definition with relation "Belongs To" should be defined the same way it is at [official gorm docs](https://gorm.io/docs/belongs_to.html#Override-References). Again, example of this can be found at [master](https://github.com/vaexey/sut-se-price-map/tree/master), specifically [here](https://github.com/vaexey/sut-se-price-map/blob/master/back/model/contrib.go).

vaexey commented on back/db/contrib.go:
What happened to whitespaces in this commit?

vaexey commented on back/api/report.go:
```diff
@@ -0,0 +1,43 @@
+	var reports []model.Report
+	var err error
+	reports, err = a.Db.Report.SelectAll()
+	if err != nil {
```
What if no records are present but database responded correctly?

vaexey reviewed (changes requested):
As stated in comments

vaexey committed:
fix region model not having id json tag

jacekrudek committed:
stores endpoint fix

fixed the stores response json structure according to doc

vaexey commented on back/api/store.go:
```diff
@@ -28,6 +27,27 @@ func (a *Api) Stores(c *gin.Context) {
+	for i := range stores {
+		region, err := a.Db.Region.SelectById(stores[i].RegionID)
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, gin.H{
```
Not consistent with OpenAPI spec as for valid response codes for this request.
*(take this feedback as a hint for future requests as this whole code region needs to be refactores regardless)*

vaexey commented on back/api/store.go:
```diff
@@ -28,6 +27,27 @@ func (a *Api) Stores(c *gin.Context) {
 	}
 
+	for i := range stores {
+		region, err := a.Db.Region.SelectById(stores[i].RegionID)
```
Fetching a request that would return `n` stores creates `n+1` (`O(n)`) database queries. Consider joining tables here instead of querying for each region (`O(1)`)

vaexey reviewed (changes requested):
Comments in files

jacekrudek committed:
minor "id endpoints" response fix

changed error responses from 400 to 404 if parameter with certain ID not found

sebasion9 commented on back/api/region.go:
```diff
@@ -54,7 +53,7 @@ func (a *Api) RegionById(c *gin.Context) {
 
 	region, err := a.Db.Region.SelectById(uint(regionID))
 	if err != nil {
```
As it is now, this will always respond with `404` on **ANY** error. Respond `404` on [ErrRecordNotFound](https://gorm.io/docs/error_handling.html#ErrRecordNotFound) else `503` (service unavailable).

jacekrudek committed:
error handling on /region/{regionId} fix

specified when errors of type 404 & 503 occur

sebasion9 reviewed (approved):
lgtm

jacekrudek committed:
region endpoint refactor

- added response structures for Region and Store to match openAPI spec
- fixed parent handling for the region endpoint
- added recursive parent counting for regions

jacekrudek committed:
adjust region & stores model structure

- changed the model for regin & stores to match both db and openAPI
- fixed the request processing accordingly to model changes

jacekrudek committed:
fix error type & region model structure

BartekWons committed:
Merge branch 'master' into feature-back

vaexey commented on back/api/store.go:
```diff
@@ -27,25 +27,26 @@ func (a *Api) Stores(c *gin.Context) {
-			return
-		}
+	// Preload the associated Region for each Store and its nested Parent regions
+	err = a.Db.Db.Preload("Region.Parent.Parent.Parent").Find(&stores).Error
```
lgetm

vaexey reviewed (commented):
lgtm

BartekWons committed:
removed unnecessary foreign keys, added check if reports table is not empty

vaexey commented on back/api/report.go:
```diff
@@ -18,6 +18,12 @@ func (a *Api) Reports(c *gin.Context) {
 	}
 
+	if len(reports) == 0 {
+		c.JSON(http.StatusNotFound, gin.H{
```
This is not the desired error response:
1. If no reports were found, this code would never be reached - `err` is not `nil` thus 503 is thrown
2. 404 should never be responded on requests returning arrays as an empty array is still a valid response

BartekWons replied:
There is no fk gorm tag there.

vaexey reviewed (changes requested):
As stated in comments

BartekWons committed:
fix edge case when no records are present in db, but db responds correctly

sebasion9 commented on back/api/report.go:
```diff
@@ -19,9 +19,7 @@ func (a *Api) Reports(c *gin.Context) {
-		c.JSON(http.StatusNotFound, gin.H{
-			"message": "Do not found any reports",
-		})
+		c.JSON(http.StatusOK, []model.Report{})
```
This does not address this [comment](https://github.com/vaexey/sut-se-price-map/pull/67#discussion_r1948101370)'s first point. This code is not reached when records are not found. Simply reference existing code at [master](https://github.com/vaexey/sut-se-price-map/tree/master), specifically [here](https://github.com/vaexey/sut-se-price-map/blob/master/back/api/contrib.go#L30).

BartekWons committed:
add filters and fix record not found error

BartekWons committed:
add skipped files

sebasion9 commented on back/db/contrib.go:
```diff
@@ -90,3 +91,9 @@ func (cs *contribService) Update(contrib model.Contrib) error {
 
+func (cs * contribService) GetNumberOfContribs(authorId uint) (int, error) {
+    var contribs []model.Contrib
+    err := cs.Db.Preload("Product").Preload("Store").Preload("Author").Where("author_id = ?", authorId).Find(&contribs).Error
```
What is the need of preloading here if only number of contribs is returned?

BartekWons replied:
Preloading was unnecessary.

sebasion9 commented on back/model/report.go:
```diff
@@ -0,0 +1,14 @@
+	Reported uint   `gorm:"column:reported_id" json:"reported"`
+	Message  string `json:"message"`
+	AuthorID uint   `json:"-"`
+	Author   User   `gorm:"column:author_id" json:"author"`
```
????? see other model definitions please, what is this column tag used for??

BartekWons replied:
fixed

sebasion9 commented on back/model/resource.go:
```diff
@@ -1,7 +1,12 @@
+	Name string
 	Url  string `json:"url"`
-	Blob string `json:"blob"`
+	Blob string `gorm:"-" json:"blob"`
```
whats the reason for `gorm"-"` here? just curious

BartekWons replied:
fixed

BartekWons committed:
fix error handling, removed unnecessary gorm tags

BartekWons committed:
code touch-ups

BartekWons committed:
db update and minior update in model files

BartekWons committed:
db update
